module/VuFind/src/VuFind/Db/Service/SearchService.php

```diff
@@ -231,16 +249,25 @@ public function getSearchesByChecksumAndOwner(
      */
     public function cleanUpInvalidUserIds(): int
     {
-        $searchTable = $this->getDbTable('search');
-        $allIds = $this->getDbTable('user')->getSql()->select()->columns(['id']);
-        $searchCallback = function ($select) use ($allIds) {
-            $select->where->isNotNull('user_id')->AND->notIn('user_id', $allIds);
-        };
-        $badRows = $searchTable->select($searchCallback);
-        $count = count($badRows);
-        if ($count > 0) {
-            $searchTable->update(['user_id' => null], $searchCallback);
+        $dql = 'SELECT u.id FROM ' . $this->getEntityClass(UserEntityInterface::class) . ' u';
+        $query = $this->entityManager->createQuery($dql);
+        $validUserIds = $query->getResult();
+        $validUserIds = array_map(fn ($user) => $user['id'], $validUserIds);
+
+        // If there are no valid users, we can skip the update
+        if (empty($validUserIds)) {
+            return 0;
         }
+
+        // Update invalid user IDs to NULL in a single query
+        $dql = 'UPDATE ' . $this->getEntityClass(SearchEntityInterface::class) . ' s '
+            . 'SET s.user = NULL '
+            . 'WHERE s.user NOT IN (:validUserIds)';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameter('validUserIds', $validUserIds);
+
+        //Number of updated records
+        $count = $query->execute();
```

Comment:
Can we simplify this to use a subquery instead of doing the work in PHP-land?

```suggestion
        // Update invalid user IDs to NULL in a single query
        $dql = 'UPDATE ' . $this->getEntityClass(SearchEntityInterface::class) . ' s '
            . 'SET s.user = NULL '
            . 'WHERE s.user NOT IN (SELECT u.id FROM ' . $this->getEntityClass(UserEntityInterface::class) . ' u)';
        $query = $this->entityManager->createQuery($dql);

        // Number of updated records
        $count = $query->execute();
```

Apologies if this syntax is not supported, but I at least hope that it might be!

Reply:
Doctrine DQL does not support subqueries in UPDATE or DELETE statements